### PR TITLE
KAFKA-3442: Fix FileMessageSet iterator.

### PR DESCRIPTION
**Pull request by becketqin**

### CODE REVIEW

File: core/src/main/scala/kafka/log/FileMessageSet.scala [hunk 1]
```diff
@@ -260,20 +261,20 @@ class FileMessageSet private[kafka](@volatile var file: File,
         sizeOffsetBuffer.rewind()
         val offset = sizeOffsetBuffer.getLong()
         val size = sizeOffsetBuffer.getInt()
-        if(size < Message.MinMessageOverhead)
+        if(size < Message.MinMessageOverhead || location + SizeOffsetLength + size > end)
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/FileMessageSet.scala]: Just to double-check: it is intentional to use `>` here even though we used `>=` a few lines above?

Comment (junrao) [inline core/src/main/scala/kafka/log/FileMessageSet.scala]: "end" is not inclusive. As long as we don't need the byte at "end" or beyond, we can stay in the iterator. So, in this case, > seems correct.

File: core/src/main/scala/kafka/log/FileMessageSet.scala [hunk 2]
```diff
@@ -245,10 +245,11 @@ class FileMessageSet private[kafka](@volatile var file: File,
   def iterator(maxMessageSize: Int): Iterator[MessageAndOffset] = {
     new IteratorTemplate[MessageAndOffset] {
       var location = start
-      val sizeOffsetBuffer = ByteBuffer.allocate(12)
+      val SizeOffsetLength = 12
```
Comment (junrao) [inline core/src/main/scala/kafka/log/FileMessageSet.scala]: Lower case Size?

**Comment by junrao** [comment]
Also, we need to properly handle the returning of a partial message when doing the down conversion. Left a comment on the jira.

File: core/src/main/scala/kafka/log/FileMessageSet.scala [hunk 3]
```diff
@@ -224,11 +224,19 @@ class FileMessageSet private[kafka](@volatile var file: File,
...
-    new ByteBufferMessageSet(
-      compressionCodec = this.headOption.map(_.message.compressionCodec).getOrElse(NoCompressionCodec),
-      offsetSeq = offsets,
-      newMessages: _*)
+    if (_size.get() > 0 && newMessages.size == 0) {
+      // This indicates that the message is too large. We just return all the bytes in the file message set.
+      val buffer = ByteBuffer.allocate(_size.get())
+      channel.read(buffer, start)
+      buffer.rewind()
+      new ByteBufferMessageSet(buffer)
```
Comment (junrao) [inline core/src/main/scala/kafka/log/FileMessageSet.scala]: Would it be better to simply return the FileMessageSet itself in this case? We will need to change toMessageFormat() to return MessageSet instead.

File: core/src/main/scala/kafka/log/FileMessageSet.scala [hunk 4]
```diff
@@ -224,11 +224,19 @@ class FileMessageSet private[kafka](@volatile var file: File,
       }
     }
 
-    // We use the offset seq to assign offsets so the offset of the messages does not change.
-    new ByteBufferMessageSet(
-      compressionCodec = this.headOption.map(_.message.compressionCodec).getOrElse(NoCompressionCodec),
-      offsetSeq = offsets,
-      newMessages: _*)
+    if (_size.get() > 0 && newMessages.size == 0) {
```
Comment (junrao) [inline core/src/main/scala/kafka/log/FileMessageSet.scala]: _size.get() => sizeInBytes Ditto below.

File: core/src/main/scala/kafka/log/FileMessageSet.scala [hunk 5]
```diff
@@ -224,11 +224,16 @@ class FileMessageSet private[kafka](@volatile var file: File,
       }
     }
 
-    // We use the offset seq to assign offsets so the offset of the messages does not change.
-    new ByteBufferMessageSet(
-      compressionCodec = this.headOption.map(_.message.compressionCodec).getOrElse(NoCompressionCodec),
-      offsetSeq = offsets,
-      newMessages: _*)
+    if (sizeInBytes > 0 && newMessages.size == 0) {
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/FileMessageSet.scala:227]: Nitpick: `newMessages.isEmpty`

**Comment by junrao** [comment]
Thanks for the patch. LGTM

**Pull request closed without merging**